File: 067__Add_Binary.py
class Solution(object):
    def addBinary(self, a, b):
        """
        :type a: str
        :type b: str
        :rtype: str
        """
        def change_binary(a):
            result = 0
            temp = 1
            while a:
                result += temp * (a % 10)
                temp *= 2
                a /= 10
            return result

        def change_hex(a):
            result = 0
            temp = 1
            while a:
                result += temp * (a % 2)
                temp *= 10
                a /= 2
            return result

        return str(change_hex(change_binary(int(a)) + change_binary(int(b))))



class Solution:
    def addBinary(self, a, b):
        if len(a) == 0: return b
        if len(b) == 0: return a
        if a[-1] == '1' and b[-1] == '1':
            return self.addBinary(self.addBinary(a[0:-1],b[0:-1]),'1')+'0'
        if a[-1] == '0' and b[-1] == '0':
            return self.addBinary(a[0:-1],b[0:-1])+'0'
        else:
            return self.addBinary(a[0:-1],b[0:-1])+'1'

File: Java/067__Add_Binary.java
class Solution {
    public String addBinary(String a, String b) {
        int i = a.length() - 1, j = b.length() - 1, incre = 0;
        String result = "";
        while (i >= 0 && j >= 0) {
            if (a.charAt(i) == '1' && b.charAt(j) == '1') {
                if (incre == 0) {
                    result = "0" + result;
                } else {
                    result = "1" + result;
                }
                incre = 1;
            } else if (a.charAt(i) == '0' && b.charAt(j) == '0') {
                if (incre == 0) {
                    result = "0" + result;
                } else {
                    result = "1" + result;
                }
                incre = 0;
            } else {
                if (incre == 0) {
                    result = "1" + result;
                } else {
                    result = "0" + result;
                }
            }
            i--;
            j--;
        }
        if (i < 0) {
            while (j >= 0) {
                if (b.charAt(j) == '0' && incre == 0) {
                    result = "0" + result;
                } else if ((b.charAt(j) == '0' && incre == 1) || (b.charAt(j) == '1' && incre == 0)) {
                    result = "1" + result;
                    incre = 0;
                } else {
                    result = "0" + result;
                    incre = 1;
                }
                j--;
            }
        } else {
            while (i >= 0) {
                if (a.charAt(i) == '0' && incre == 0) {
                    result = "0" + result;
                } else if ((a.charAt(i) == '0' && incre == 1) || (a.charAt(i) == '1' && incre == 0)) {
                    result = '1' + result;
                    incre = 0;
                } else {
                    result = "0" + result;
                    incre = 1;
                }
                i--;
            }
        }
        if (incre == 1) {
            result = "1" + result;
        }
        return result;
    }
}
